src/peft/tuners/lora/tp_layer.py

```diff
@@ -0,0 +1,145 @@
+from typing import Any
+
+import torch
+import torch.nn as nn
+import torch.nn.init as init
+
+from .layer import LoraLayer
+
+
+class LoraParallelLinear(nn.Module, LoraLayer):
+    """
+    When the target layer parallel_linear is RowParallelLinear, in order to keep the input and output shapes
+    consistent, we need to split the lora matrix A into rows, and the lora_B at this time should be a complete linear
+    layer; In the same way, when the target layer is ColumnParallelLinear, we perform column segmentation on lora_B,
+    while lora_A is still a complete linear layer.
+    """
+
+    def __init__(
+        self,
+        base_layer,
+        adapter_name: str,
+        backend,
+        r: int = 0,
+        lora_alpha: int = 1,
+        lora_dropout: float = 0.0,
+        fan_in_fan_out: bool = False,
+        init_lora_weights: bool = True,
+        **kwargs,
+    ):
+        super().__init__()
+        LoraLayer.__init__(self, base_layer=base_layer)
+
+        self.backend = backend
+        self.is_paralle_a = isinstance(base_layer, backend.RowParallelLinear)
+        self.fan_in_fan_out = fan_in_fan_out
+        self._active_adapter = adapter_name
+
+        megatron_config = kwargs["megatron_config"]
+        parallel_linear_kwargs = {"megatron_config": megatron_config}
```

Comment:
If `megatron_config` is `None`, would this still work? Should we put a default `TransformerConfig` here instead?

Reply:
> If `megatron_config` is `None`, would this still work? Should we put a default `TransformerConfig` here instead?

ok~

Reply:
> If `megatron_config` is `None`, would this still work? Should we put a default `TransformerConfig` here instead?

Because in model.py, we will judge whether megatron_config has value, and if so, we will import megatron_core. Only when megatron_core is successfully imported will it go to tp_layer.py.
Therefore, the megatron_config here can work. And because we don’t know the configured environment information, we can’t give a default TransformerConfig.

Reply:
I see, thanks